src/drivers/intel/ssp/mn.c

```diff
@@ -569,10 +569,16 @@ int mn_set_bclk(uint32_t dai_index, uint32_t bclk_rate,
 void mn_release_bclk(uint32_t dai_index)
 {
 	struct mn *mn = mn_get();
+	bool mn_in_use;
 
 	spin_lock(&mn->lock);
 	mn->bclk_sources[dai_index] = MN_BCLK_SOURCE_NONE;
 	platform_shared_commit(mn, sizeof(*mn));
+
+	mn_in_use = is_bclk_source_in_use(MN_BCLK_SOURCE_MN);
```

Comment:
I think that's incorrect. There are platforms where we can have different sources for BCLK and MCLK, but other where there's a single source.

Reply:
@plbossart but do we use the MN for MCLK too?

Reply:
M/N comes *after* selecting a source, and yes it's possible to apply it there as well IIRC.

Reply:
@plbossart we have MNDSS and MCDSS for all cAVS, at least I am seeing them in spec for APL/CNL/TGL. This driver is for cAVS platforms only, so no need to take care BYT/BDW ones here.

@ranj063 yes, please see the definition and block diagram for MNDSS(for BCLK) and MCDSS(for MCLK). We have M/N for BCLK, but only divider (similar to N) for MCLK, which means we can only set the MCLK frequency as 1/N of the MCLK source frequency.

Reply:
@keyonjie I am 100% positive that there was a change in the clock tree somewhere between APL and TGL. In fact here it is:

see commit 0a97c1a92f2d93bd4d45bc99d61e362cd214748c

````
    drivers: intel: mn: bclk sources
    
    All BCLKs that use M/N share the same source clock, so we need to
    track sources for each BCLK to prevent case in which next SSP config
    would override source clock for previous SSP config.
    
    Since cAVS 2.0, all clocks go through M/N, bypassing XTAL is
    no longer supported.
    For earlier version MN driver checks if XTAL can be used if it
    would result in more efficient settings.
    As a result it fixes ECS-related code for cAVS 2.0+ and is backwards
    compatible with previous logic when there is only SSP0 used.
    Previously topologies that used more than 1 SSP could override
    source clock for previously configured SSP ports, now it will try
    to use current clock with M/N or report error.
````

Reply:
@plbossart Yes, M/N can not be bypassed on cAVS2.x but we can set both M and N to 1.